Add androidNative targets.

MohammedKHC opened the pull request:
This adds support for androidNative targets.

This is the second attempt at trying to do this. Now I can confirm that it does compile and work.

The problem with the first attempt (#1726) was that it didn't account for the fact that platform.posix.DIR doesn't exist on androidNative. Therefore, opendir, readdir, and closedir can't be used from common code. This fixes this by using expect/actual for all this stuff.

To run the tests. Run the following commands in order.
(You need to have an Android Emulator running first.)
```
./gradlew androidNativeX64TestBinaries
adb push /path/to/okio/okio/build/bin/androidNativeX64/debugTest/test.kexe /data/local/tmp
rm -r /path/to/okio/build
adb push /path/to/okio /data/local/tmp/okio
adb shell
export OKIO_ROOT=/data/local/tmp/okio
/data/local/tmp/test.kexe
```

Closes: #1692, Closes #1242

MohammedKHC committed:
Add androidNative targets.

MohammadKHC commented:
UPDATE: I got most tests running.
Those are the only failing tests.
```
[  FAILED  ] 3 tests, listed below:
[  FAILED  ] okio.NativeSystemFileSystemTest.canonicalizeDotReturnsCurrentWorkingDirectory
[  FAILED  ] okio.NativeSystemFileSystemTest.listOnRelativePathWhichIsNotDotReturnsRelativePaths
[  FAILED  ] okio.NativeSystemFileSystemTest.listOrNullOnRelativePathWhichIsNotDotReturnsRelativePaths
```

To run tests now, do the following on the terminal.
```
./gradlew androidNativeX64TestBinaries
adb push /path/to/okio/okio/build/bin/androidNativeX64/debugTest/test.kexe /data/local/tmp
rm -r /path/to/okio/build
adb push /path/to/okio /data/local/tmp/okio
adb shell
cd /data/local/tmp
export OKIO_ROOT=/data/local/tmp/okio
./test.kexe
```

MohammadKHC commented:
UPDATE again.
`canonicalizeDotReturnsCurrentWorkingDirectory` fails because i changed the cwd to /data/local/tmp..
The test expects it to be "/". And it does not fail if i don't do the cd.

Now the only failling tests are `listOnRelativePathWhichIsNotDotReturnsRelativePaths` and `listOrNullOnRelativePathWhichIsNotDotReturnsRelativePaths`

Best way to run the tests as of now.
```
./gradlew androidNativeX64TestBinaries
adb push /path/to/okio/okio/build/bin/androidNativeX64/debugTest/test.kexe /data/local/tmp
rm -r /path/to/okio/build
adb push /path/to/okio /data/local/tmp/okio
adb shell
export OKIO_ROOT=/data/local/tmp/okio
/data/local/tmp/test.kexe
```

MohammadKHC commented:
After reading the `listOnRelativePathWhichIsNotDotReturnsRelativePaths` code, it seems that the test itself is flaky. It only creates the required `okio.api` test file when one of the following is true:
`isFakeFileSystem || isWrappingJimFileSystem || isWasiFileSystem`

Knowing that, I can say that Okio does work as intended on Android native, which is useful for a bunch of Android tech apps such as Termux or Android IDEs.

swankjesse commented on okio/build.gradle.kts:
```diff
@@ -28,6 +28,11 @@ plugins {
  *   |   |-- mingw
  *   |   |   '-- mingwX64
  *   |   '-- unix
+ *   |       |-- androidNative
```
I would have expected this under linux

swankjesse reviewed (commented):
Looks very promising. Thanks for driving this.

I wanna look into making androidNativeMain a child of linuxMain

MohammedKHC committed:
Make Android Native a child of Linux.
and workaround `platform.posix.DIR` not being available there.

MohammedKHC committed:
Add Android Native support for opendir/readdir/closedir.

MohammedKHC committed:
Add `PosixDirectory` and use it instead of expect/actual `DIR` functions.

Unfortunately we cannot typealias `DIR` in a shared source set. Because it is already a typealias on Linux.

JakeWharton commented on okio/src/nativeMain/kotlin/okio/internal/PosixDirectory.kt:
```diff
@@ -0,0 +1,34 @@
+internal expect class PosixDirectory(path: Path): Closeable {
+  val isInvalid: Boolean
+  fun nextEntry(): CPointer<dirent>?
+}
```
Seems like these should all be `value class`es to avoid the memory allocation. Does anything prevent that?

MohammadKHC replied:
@JakeWharton 
The problem is that the actual implementation needs to store
```kotlin
private val dir = opendir(path.toString())
```
And if you meant that the value should be DIR. Then this is kinda complicated.
as we cannot use expect/actual for DIR in only two source sets. (androidNative, nativeNonAndroid)

that's because this code will not compile under nativeNonAndroid
```kotlin
actual typealias DIR = platform.posix.DIR
```
It will fail with an error cannot create a typealias to a typealias.
As platform.posix.DIR is a typealias on Linux.

we could create a new nativeNonLinux source set that will exclude androidNative and linux.

What do you think?

MohammedKHC committed:
Use value class for `PosixDirectory`.

MohammedKHC committed:
Merge branch 'master' of github.com:square/okio into android-native

MohammedKHC committed:
Use statx on Android native.

MohammedKHC committed:
Inline openPosixDirectory

sky130 commented:
Hi @MohammedKHC,
I tried building your branch on my local device, but it failed. I looked at the source code and found that functions on the arm32 architecture accept int type, while functions on the arm64 architecture accept long type. Do you have time to investigate this issue? If not, I will try to resolve it later.

---
source code

<img width="2218" height="1298" alt="image" src="https://github.com/user-attachments/assets/cd5eaaca-1811-419e-88b0-8cd4d86f5b21" />

---

``` log
> Task :okio:compileKotlinAndroidNativeArm32 FAILED
e: file:///Users/***/Desktop/projects/okio/okio/src/linuxMain/kotlin/okio/LinuxPosixVariant.kt:47:7 Argument type mismatch: actual type is 'Long', but 'Int' was expected.
e: file:///Users/***/Desktop/projects/okio/okio/src/linuxMain/kotlin/okio/LinuxPosixVariant.kt:54:9 Operator '==' cannot be applied to 'Int' and 'Long'.

Execution failed for task ':okio:compileKotlinAndroidNativeArm32'.
> Compilation finished with errors
```

MohammedKHC committed:
Merge branch 'master' into android-native

MohammedKHC committed:
Fix statx syscall for arm32

MohammedKHC commented:
Hi! @sky130 
Thanks for catching this! I fixed it just now. https://github.com/square/okio/pull/1739/commits/dfef41683a7a23c63bb88d277323a2a873831118
I forgot to compile to arm32 after the new statx implementation got merged, sorry!

MohammedKHC commented:
@JakeWharton can this be merged anytime soon?
I almost forgot about it.

sky130 commented:
Hi @MohammedKHC ,

I meet another problem now. The app crashes with a **SIGSYS signal** on Android API 26(AVD).

Stack trace points to `okio.PosixFileSystem.variantMetadataOrNull()` calling the `statx` syscall. On  Android API 26, `statx` is blocked by `seccomp` and the kernel **directly sends SIGSYS** to the process. Execution never reaches the fallback `lstat` path because the signal terminates the app immediately.

Do you have time to look into this issue? If you don't have time, I'll try to fix it.

```
Fatal signal 31 (SIGSYS), code 1 in tid 29348 (package-name)
(kfun:okio#variantMetadataOrNull__at__okio.PosixFileSystem(okio.Path){}okio.FileMetadata?+1136)
(kfun:okio.PosixFileSystem#metadataOrNull(okio.Path){}okio.FileMetadata?+88)
(kfun:okio.FileSystem#metadataOrNull(okio.Path){}okio.FileMetadata?-trampoline+64)
(kfun:okio.internal#commonExists__at__okio.FileSystem(okio.Path){}kotlin.Boolean+132)
(kfun:okio.FileSystem#exists(okio.Path){}kotlin.Boolean+80)
(kfun:okio.internal#commonCreateDirectories__at__okio.FileSystem(okio.Path;kotlin.Boolean){}+536)
(kfun:okio.FileSystem#createDirectories(okio.Path;kotlin.Boolean){}+108)
(kfun:okio.FileSystem#createDirectories$default(okio.Path;kotlin.Boolean;kotlin.Int){}+172)
```

MohammedKHC committed:
Merge branch 'master' into android-native

swankjesse reviewed (commented):
Sorry for sitting on this for such a long time.

I think this is good stuff. My only request is to change `PosixDirectory` from an `expect class` to an interface, that'll have different implementations on each platform.

ALSO – I’d prefer to be able to run some kind of androidNative test cases in CI. Any idea what’s involved in that?

MohammedKHC committed:
Merge branch 'main' into android-native

MohammedKHC committed:
Make `PosixDirectory` an interface.

MohammedKHC committed:
Fix statx on Android native.

MohammedKHC committed:
Cleanup

MohammedKHC committed:
apiDump

MohammedKHC committed:
Test android native on CI

MohammedKHC committed:
Set `fileSystemHasGoodMetadata` to false on Android.

And ignore the following tests
1. `listOnRelativePathWhichIsNotDotReturnsRelativePaths`
2. `listOrNullOnRelativePathWhichIsNotDotReturnsRelativePaths`

as we don't really want to change the working directroy of the running test executable.

MohammedKHC committed:
Merge branch 'main' into android-native

MohammedKHC committed:
apiDumb fakefilesystem.

MohammedKHC committed:
Fix browser tests.
